tests/fips/fips_setup.pm

```diff
@@ -13,43 +13,71 @@
 # with this program; if not, see <http://www.gnu.org/licenses/>.
 #
-# Summary: Setup fips mode for further testing
-#          Support both "global mode" (fips=1 in kernel command line)
-#          and "ENV mode" - selected by FIPS_ENV_MODE
-# Maintainer: Wes <whdu@suse.com>
+# Summary: Setup fips mode for further testing:
+#          Installation check - verify the setup of FIPS installation
+#          ENV mode - selected by FIPS_ENV_MODE
+#          Global mode - setup fips=1 in kernel command line
+# Maintainer: wnereiz <wnereiz@member.fsf.org>
 # Tags: poo#39071
 
 use strict;
 use warnings;
 use base "consoletest";
 use testapi;
 use utils 'zypper_call';
-use Utils::Systemd 'disable_and_stop_service';
+use bootloader_setup 'add_grub_cmdline_settings';
+use power_action_utils "power_action";
 
 sub run {
+    my ($self) = @_;
     select_console 'root-console';
 
-    disable_and_stop_service('packagekit.service', mask_service => 1);
+    # For installation only. FIPS has already been setup during installation
+    # (DVD installer booted with fips=1), so we only do verification here.
+    if (get_var("FIPS_INSTALLATION")) {
+        assert_script_run("grep '^GRUB_CMDLINE_LINUX_DEFAULT.*fips=1' /etc/default/grub");
+        assert_script_run("grep '^1\$' /proc/sys/crypto/fips_enabled");
+        record_info 'Kernel Mode', 'FIPS kernel mode (for global) configured!';
 
-    zypper_call('in -t pattern fips');
+        # Make sure FIPS pattern is installed and there is no conflicts.
+        zypper_call("refresh");
+        zypper_call("search -si -t pattern fips");
+
+        return;
+    }
+
+    # FIPS_INSTALLATION is only applicable for system installaton
+    die "FIPS_INSTALLATION is require to run this script for installation" if get_var("!BOOT_HDD_IMAGE");
+    die "FIPS setup is only applicable for FIPS_ENABLED=1 image!"          if get_var("!FIPS_ENABLED");
 
     # If FIPS_ENV_MODE, then set ENV for some FIPS modules. It is a
     # workaround when fips=1 kernel cmdline is not working.
     # If FIPS_ENV_MODE does not set, global FIPS mode (fips=1 from
     # kernel command line) will be applied
     if (get_var("FIPS_ENV_MODE")) {
+        die 'FIPS kernel mode is required for this test!' if check_var('SECURITY_TEST', 'crypt_kernel');
+        zypper_call('in -t pattern fips');
         foreach my $env ('OPENSSL_FIPS', 'OPENSSL_FORCE_FIPS_MODE', 'LIBGCRYPT_FORCE_FIPS_MODE', 'NSS_FIPS') {
             assert_script_run "echo 'export $env=1' >> /etc/bash.bashrc";
         }
+
+        record_info 'ENV Mode', 'FIPS environment mode (for single modules) configured!';
     }
     else {
-        assert_script_run "sed -i '/^GRUB_CMDLINE_LINUX_DEFAULT/s/\\(\"\\)\$/ fips=1\\1/' /etc/default/grub";
-        assert_script_run "grub2-mkconfig -o /boot/grub2/grub.cfg";
+        zypper_call('in -t pattern fips');
+        add_grub_cmdline_settings('fips=1', 1);
+        record_info 'Kernel Mode', 'FIPS kernel mode configured!';
     }
+
```

Comment:
Line 67 and line 59  are same (zypper_call('in -t pattern fips');), so can be moved to here

Reply:
Two `zypper_call`s are in different position in the code for two conditions, so it is not possible to move it out of if conditions. And just one line of duplicate is not a big problem I suppose.